### PR TITLE
Added more testing and idiom to show how to get the single Boss

### PR DESCRIPTION
**Pull request by AndyDentFree**
Want to merge these tests in, with `RecursiveBacklinksQueryFailures` disabled as Explicit, so that can be just enabled for the fix later.

### CODE REVIEW

File: Shared/Tests.Shared/TestObjects.cs [hunk 1]
```diff
@@ -209,6 +210,25 @@ public class Owner : RealmObject
...
+        public IList<Employee> Reports { get; }
+
+        [Backlink(nameof(Employee.Reports))]
+        public IQueryable<Employee> Bosses { get; }
+
+        public Employee Boss
+        {
+            get
+            {
+                return Bosses.ToList().SingleOrDefault();
```
Comment (fealebenpae) [inline Shared/Tests.Shared/TestObjects.cs]: Shouldn't this be just `Bosses.SingleOrDefault()` to trigger the error condition?

Comment (AndyDentFree) [inline Shared/Tests.Shared/TestObjects.cs]: I wrote it that way to show a workaround for the period up until it is fixed. Would you like a comment on there to make that clear (sorry I should have had one)?

File: Shared/Tests.Shared/RelationshipTests.cs [hunk 2]
```diff
@@ -437,6 +437,56 @@ public void Backlinks()
             Assert.That(maggie.Owners, Is.EquivalentTo(new[] { dani }));
         }
 
+        [Test]
+        public void RecursiveBacklinksBossTestOnly()
```
Comment (kristiandupont) [inline Shared/Tests.Shared/RelationshipTests.cs]: I'd prefer if we could try and stick with a more spec-style naming of our tests. The fact that this involves a boss is not relevant. This test could be named `BacklinksShouldWorkRecursively`

File: Shared/Tests.Shared/RelationshipTests.cs [hunk 3]
```diff
@@ -437,6 +437,56 @@ public void Backlinks()
...
+
+            // simple query of immediate Boss 
+            var jake = realm.All<Employee>().Single(p => p.Name == "Jake");
+            Assert.That(jake.Boss?.Name, Is.EqualTo("Sally"));
+            var bossesBoss = jake.Boss?.Boss;
+            Assert.That(bossesBoss, Is.Null);
+        }
+
+        // Demonstrate the bug reported in issue 1177
+        [Test, NUnit.Framework.Explicit("Demonstrates issue 1177 so disable from auto run")]
```
Comment (kristiandupont) [inline Shared/Tests.Shared/RelationshipTests.cs]: Do we need to specify the namespace here?

File: Shared/Tests.Shared/TestObjects.cs (hunk 1 above)
Comment (nirinchev) [inline Shared/Tests.Shared/TestObjects.cs]: Why do our tests need to show a workaround?

File: Shared/Tests.Shared/RelationshipTests.cs (hunk 3 above)
Comment (fealebenpae) [inline Shared/Tests.Shared/RelationshipTests.cs]: Yes, because we have the internal `Realms.ExplicitAttribute` and they conflict.

Comment (AndyDentFree) [inline Shared/Tests.Shared/RelationshipTests.cs]: Yes, it wasn't building until I did that.

File: Shared/Tests.Shared/TestObjects.cs (hunk 1 above)
Comment (AndyDentFree) [inline Shared/Tests.Shared/TestObjects.cs]: It was unclear if we are going to ship a fix before the next release - as there's a simple but not necessarily obvious workaround I wanted to include it.

Comment (nirinchev) [inline Shared/Tests.Shared/TestObjects.cs]: I am quite opposed to using tests as documentation. Putting myself in users' shoes, if I get weird behavior, I wouldn't start browsing unit tests, I'll check the docs and if there's nothing in there, I'd open an issue.

Comment (hrach) [inline Shared/Tests.Shared/TestObjects.cs]: @nirinchev I'm just novice user to c# and unit test were the first I'd go for when I didn't google any info. #1084 I have no strong opinion on this, I want just add that users may take a look at tests when they are desperate. Generally, 👍  for putting info about workarounds to docs!

**Comment by nirinchev** [comment]
This is superseded by #1206 for backlink issues. AFAICT the other tests were used for debugging purposes only and don't demonstrate the problematic behavior in a testable manner. I suggest extracting the gist of these in a new issue to improve our query provider performance and closing this PR.

**Comment by AndyDentFree** [comment]
Relevant test has been grabbed, this is no longer needed.

**Pull request closed without merging**